src/OrleansRuntime/Core/Dispatcher.cs

```diff
@@ -597,32 +598,60 @@ private void ResendMessageImpl(Message message, ActivationAddress forwardingAddr
         #region Send path
 
         /// <summary>
-        /// Send an outgoing message
+        /// Send an outgoing message, may complete synchronously
         /// - may buffer for transaction completion / commit if it ends a transaction
         /// - choose target placement address, maintaining send order
         /// - add ordering info and maintain send order
         /// 
         /// </summary>
         /// <param name="message"></param>
         /// <param name="sendingActivation"></param>
-        public async Task AsyncSendMessage(Message message, ActivationData sendingActivation = null)
+        public Task AsyncSendMessage(Message message, ActivationData sendingActivation = null)
         {
-            try
-            {
-                await AddressMessage(message);
-                TransportMessage(message);
-            }
-            catch (Exception ex)
+            Action<Exception> onAddressingFailure = ex =>
             {
                 if (ShouldLogError(ex))
                 {
-                    logger.Error(ErrorCode.Dispatcher_SelectTarget_Failed,
-                        String.Format("SelectTarget failed with {0}", ex.Message),
-                        ex);
+                    logger.Error(ErrorCode.Dispatcher_SelectTarget_Failed, $"SelectTarget failed with {ex.Message}", ex);
                 }
+
                 MessagingProcessingStatisticsGroup.OnDispatcherMessageProcessedError(message, "SelectTarget failed");
                 RejectMessage(message, Message.RejectionTypes.Unrecoverable, ex);
+            };
+
+            Func<Task, Task> transportMessageAfterAddressing = async addressMessageTask =>
```

Comment:
Initially they were inline methods, but it's not yet supported in Orleans, so converted to ordinary lambdas. While they will allocate on each call - that objects will end up collected in gen 0,  quite unlikely will affect benchmarks\real world usages, and I do believe, that such kind of methods is better from support perspective, as they do not pollute the internal api of class. And the codebase maybe will be upgraded to support such methods before next release.